### PR TITLE
[ENG-1257] define: CatalogVariable type for substitution

### PR DESCRIPTION
**Pull request by Cobalt0s**
# Changes

Catalog substitutions were changed from `map` to `concrete type`.

We already have connector options defined under `paramsbuilder` package.
Parameters that are of CatalogVarialbe type can be directly passed into variadic `ReadInfo` function.

Any new parameters that will participate in substitution algorithm must return SubstitutionPlan. For example `paramsbuilder.Workspace` returns SubstitutionPlan{from:"workspace", to: "DesiredValue"}.

Also, sometimes ReadInfo or ReadCatalog perform adjustment of default catalog beforehand. This is now reflected in CustomizedCatalog(... CatalogOption). This allows stacking of 2 variadic configurations of CatalogOption and CatalogVariable.

# Improvements

We could clean up providers/utils.go even further. For example clone method is very general in nature and can be used anywhere by anybody. And algorithm to `substituteStruct` can live in dedicated package.

### CODE REVIEW

**Comment by Cobalt0s** [comment]
@RajatPawar I changed the README. Let me know if it is ok.
Also, can clients implement CatalogVariable? Or is it only us that enforce allowed variables (workspace, region, etc).
The build is failing but the changes are expected to be breaking: moving from map to typed variadic methods.

File: providers/utils.go [hunk 1]
```diff
@@ -37,33 +39,59 @@ func WithCatalog(c CatalogType) CatalogOption {
...
 }
 
-func ReadCatalog(opts ...CatalogOption) (CatalogType, error) {
+type CustomCatalog struct {
+	custom CatalogType
+}
+
+// CustomizedCatalog allows to apply modifiers on the base catalog, to tweak its content.
+// Just like the default catalog it supports reading data, resolves variable substitutions.
+func CustomizedCatalog(opts ...CatalogOption) CustomCatalog {
```
Comment (RajatPawar) [inline providers/utils.go:48]:
```suggestion
func NewCustomCatalog(opts ...CatalogOption) CustomCatalog {
```

**Review by RajatPawar — COMMENTED** [review]
Looks great, apart from a couple of minor comments. Would also be great if you could test these new methods out before merging it, especially since these touch major components like reading a catalog, and substitution of values. This can be via a test script or manual. Feel free to re-request a review as soon as you are done.

File: common/paramsbuilder/params.go [hunk 2]
```diff
@@ -79,8 +79,11 @@ func (p *Workspace) WithWorkspace(workspaceRef string) {
 	p.Name = workspaceRef
 }
 
-func (p *Workspace) Substitution() map[string]string {
-	return map[string]string{"workspace": p.Name}
+func (p *Workspace) getSubstitution() SubstitutionPlan {
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply common/paramsbuilder/params.go:82]: I wasn't sure which name to stick to. I will be more explicit.

**Comment by RajatPawar** [comment]
FYI, the build and test action is failing

File: common/paramsbuilder/variables.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	getSubstitution() SubstitutionPlan
+}
+
+// SubstitutionPlan tells what string should be replaced with what value.
+type SubstitutionPlan struct {
+	from string
+	to   string
+}
+
+type SubstitutionRegistry map[string]string
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply common/paramsbuilder/variables.go]: Ok, I am not very experienced with Golang generics more with Java.
I managed to achieve generics a bit different here.

Tell me if it is ok: 
https://github.com/amp-labs/connectors/blob/fd60076223b2bd0d32abcc1aa756ffe359d5e9d6/common/paramsbuilder/variables.go#L24-L39

**Review by RajatPawar — APPROVED** [review]

**Pull request merged**